source/common/quic/quic_transport_socket_factory.h

```diff
@@ -11,44 +11,91 @@
 public:
+  QuicTransportSocketFactoryBase(Stats::Scope& store, std::unique_ptr<Ssl::ContextConfig> config,
+                                 const std::string& perspective)
+      : stats_(generateStats(store, perspective)), context_config_(std::move(config)) {
+    context_config_->setSecretUpdateCallback([this]() {
+      // The callback also triggers updating |context_config_| with the new secret.
+      // TODO(danzh) Client transport socket factory may also need to update quic crypto.
+      stats_.context_config_update_by_sds_.inc();
+    });
+  }
+
   // Network::TransportSocketFactory
   Network::TransportSocketPtr
   createTransportSocket(Network::TransportSocketOptionsSharedPtr /*options*/) const override {
     NOT_REACHED_GCOVR_EXCL_LINE;
   }
   bool implementsSecureTransport() const override { return true; }
   bool usesProxyProtocolOptions() const override { return false; }
+
+protected:
+  const Ssl::ContextConfig& contextConfig() const { return *context_config_; }
+  QuicTransportSocketFactoryStats stats_;
+
+private:
+  std::unique_ptr<Ssl::ContextConfig> context_config_;
 };
 
 // TODO(danzh): when implement ProofSource, examine of it's necessary to
 // differentiate server and client side context config.
 class QuicServerTransportSocketFactory : public QuicTransportSocketFactoryBase {
 public:
-  QuicServerTransportSocketFactory(Ssl::ServerContextConfigPtr config)
-      : config_(std::move(config)) {}
-
-  const Ssl::ServerContextConfig& serverContextConfig() const { return *config_; }
+  QuicServerTransportSocketFactory(Stats::Scope& store, Ssl::ServerContextConfigPtr config)
+      : QuicTransportSocketFactoryBase(store, std::move(config), "server"),
+        config_(dynamic_cast<const Ssl::ServerContextConfig&>(contextConfig())) {}
+
+  // Return TLS certificates if the context config is ready.
+  std::vector<std::reference_wrapper<const Envoy::Ssl::TlsCertificateConfig>>
+  getTlsCertificates() const {
+    if (!config_.isReady()) {
+      ENVOY_LOG(warn, "SDS hasn't finished updating Ssl context config yet.");
+      stats_.downstream_context_secrets_not_ready_.inc();
+      return {};
+    }
+    return config_.tlsCertificates();
+  }
 
 private:
-  std::unique_ptr<const Ssl::ServerContextConfig> config_;
+  const Ssl::ServerContextConfig& config_;
 };
 
 class QuicClientTransportSocketFactory : public QuicTransportSocketFactoryBase {
 public:
-  QuicClientTransportSocketFactory(Envoy::Ssl::ClientContextConfigPtr config)
-      : config_(std::move(config)) {}
+  QuicClientTransportSocketFactory(Stats::Scope& store, Ssl::ClientContextConfigPtr config)
+      : QuicTransportSocketFactoryBase(store, std::move(config), "client"),
+        config_(dynamic_cast<const Ssl::ClientContextConfig&>(contextConfig())) {}
 
-  const Ssl::ClientContextConfig& clientContextConfig() const { return *config_; }
+  const Ssl::ClientContextConfig& clientContextConfig() const { return config_; }
```

Comment:
I think an alternative way to do this would be to have this getter do 

````
ASSERT(dynamic_cast<...> != nullptr);
return static_cast<...>(contextConfig());
```

That way we avoid storing a separate ref on the derived classes while still avoiding a dynamic cast in prod builds?

Reply:
see above.